snuba/consumers/consumer_builder.py

```diff
@@ -202,6 +208,10 @@ def __build_streaming_strategy_factory(
                     self.producer if self.replacements_topic is not None else None
                 ),
                 replacements_topic=self.replacements_topic,
+            )
+            if not self.__mock
+            else build_mock_batch_writer(
+                self.storage, bool(self.replacements_topic), self.metrics
```

Comment:
This isn't actionable for this PR but I do want to raise something here that I've seen throughout the codebase. 

We have a lot of `Factory` classes and `build_something` functions and that makes it really hard to follow the code unless you're running it in the debugger. We have a big pattern of injecting things that are functions (or look a lot like functions) into other things that wrap them and return other functions. I understand the desire for levels of indirection (as it makes things theoretically extensible) but the ability to understand the code as written goes down.  

Again, not actionable but since I saw this and went down a rabbit hole to understand what these things did it made me remember this again. I don't know what the answer is yet except maybe a more unifying abstraction but this is something that's often in the back of my head.

Reply:
I think there are multiple concerns that you are raising:
- factories and builders. (objects instantiated whoes only goal is to instantiate another object)
- using functions as first class concepts and passing them around. 
The concerns produced by each and the reason for using vs not using them are very different IMO, and they are not just about adding indirection. 

I don't think there is anything wrong with the concept of either, especially passing around functions is a well established functional programming pattern and can produce very readable and elegant code. The issue can rise in the individual use case, if too much indirection is added. 
So I would not claim that these practices are either wrong or right in absolute in our code base.

In the specific use case. Building a `ProcessingStrategy` for the consumer to use it, the need of a factory is mandated by the fact that the `Consumer` needs to be able to create, destroy and recreate a `ProcessingStrategy` (which is a sequence of processing steps to simplify) when it needs. Examples are during partition assignment and revoking. 
We separated the `Consumer` concern (which infrastructure and specific to how Kafka works) from the processing concern (the `ProcessingStrategy`, which is application specific). I would argue the idea is good as we touch the consumer much less than we touch the processing logic. Though it has the downside that the consumer needs to be able to create and destroy the processing logic when it wants, thus the need of a factory there. 
Can this be organized in a way that makes the processing strategy code more readable. Probably. Never tried. If you have an idea, please feel free to file a backlog task so we can move ahead on that.